File: LeetCode/901-1000/994. 腐烂的橘子.java
class Solution {
        static int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public int orangesRotting(int[][] grid) {
     Queue<int[]> queue = new LinkedList<>();
        int Row = grid.length,Col = grid[0].length;
        int res = 0,count = 0;
        boolean[][] visited = new boolean[Row][Col];
        for (int i = 0; i < Row; i++) {
            for (int j = 0; j < Col; j++) {
                if (grid[i][j] == 2) {
                    queue.add(new int[]{i,j});
                    visited[i][j] = true;
                } else if (grid[i][j] == 1)count++;
            }
        }
        while (!queue.isEmpty()) {
            int n = queue.size();
            res++;
            for (int i = 0; i < n; i++) {
                int[] cur = queue.poll();
                for (int j = 0; j < 4; j++) {
                    int nr = cur[0] + dirs[j][0];
                    int nc = cur[1] + dirs[j][1];
                    if (nr >= 0 && nr < Row && nc >= 0 && nc < Col && !visited[nr][nc] && grid[nr][nc] == 1) {
                        queue.add(new int[]{nr,nc});
                        visited[nr][nc] = true;
                        count--;
                    }
                }   
            }
        }
 if (count == 0) return Math.max(0,res - 1);
        return -1;
    }
}

File: LeetCode/901-1000/994. 腐烂的橘子.cpp
class Solution {
 int count;
    int direction[4][2] = {
        {-1,0},
        {1,0},
        {0,-1},
        {0,1}
    };
    int pos[10][10];
public:
    int orangesRotting(vector<vector<int>>& grid) {
        memset(pos, -1, sizeof(pos));
        queue<pair<int, int>> q;
        count = 0;
        int m = grid.size(), n = grid[0].size();
        int ans = 0;
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (grid[i][j] == 2) {
                    q.push(make_pair(i, j));
                    pos[i][j] = 0;
                }
                else if (grid[i][j] == 1) {
                    count++;
                }
            }
        }
        while (!q.empty())
        {
            pair<int,int> f = q.front();
            q.pop();
            for (int i = 0; i < 4; i++)
            {
                int nextX = direction[i][0] + f.first;
                int nextY = direction[i][1] + f.second;
                if (nextX < 0 || nextX >= m || nextY < 0 || nextY >= n ||
                    ~pos[nextX][nextY] || !grid[nextX][nextY]) continue;
                pos[nextX][nextY] = pos[f.first][f.second] + 1;
                q.push(make_pair(nextX, nextY));
                if (grid[nextX][nextY] == 1) {
                    count--;
                    ans = pos[nextX][nextY];
                    if (!count) break;
                }
            }
        }
        return count ? -1 : ans; 
    }
};